packages/transition-frontend/src/services/definitions/DefinitionsService.ts

```diff
@@ -5,23 +5,49 @@
  * License text available at https://opensource.org/licenses/MIT
  */
 
-import React from 'react';
 import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
 import * as Status from 'chaire-lib-common/lib/utils/Status';
 import { Dictionary } from 'lodash';
 
+const definitionCache: { [label: string]: Dictionary<string> | 'loading' | 'error' } = {};
+
 export const getDefinitionFromServer = (
     label: string,
-    setDefinition: React.Dispatch<React.SetStateAction<Dictionary<string>>>,
-    setGotError: React.Dispatch<React.SetStateAction<boolean>>
+    setDefinition: (arg: Dictionary<string>) => void,
+    setGotError: (arg: boolean) => void
 ) => {
+    // If the cache is loading or already set, return the definition or wait for the completion
+    if (definitionCache[label] !== undefined && definitionCache[label] !== 'error') {
+        // If the definition is already loaded, return it
+        if (definitionCache[label] === 'loading') {
+            // Wait for the previous call to complete
+            const interval = setInterval(() => {
+                if (definitionCache[label] !== 'loading') {
+                    clearInterval(interval);
+                    if (definitionCache[label] === 'error') {
+                        setGotError(true);
+                    } else {
+                        setDefinition(definitionCache[label] as Dictionary<string>);
+                    }
+                }
+            }, 500);
+        } else {
+            setDefinition(definitionCache[label] as Dictionary<string>);
+        }
+        return;
```

Comment:
instead of return here, maybe we could simply have an else for the rest? The nesting is not too deep in the function.